[RNMobile]: Fix search block focus issues

AmandaRiu opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->
**NOTE: ~[This PR](https://github.com/WordPress/gutenberg/pull/29275) must be merged before this one can be taken out of draft~**

**Gutenberg Mobile PR:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3210

## Description
<!-- Please describe what you have changed or added -->
Fixes an couples issues found in a previous PR review:
- iOS: When adding a new search block when there are other text-based blocks already on the page would cause a bizarre UI issue where the search block kept stealing the focus causing a strange blinking of focusing between the search block and the other text blocks.
- Also on iOS when adding the search block, the app would appear to hang for a while before finally adding the block. 
- Clicking on the placeholder text input when the search block was not selected would not select the block. 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Tested on both Android emulator and iOS simulator
1. Add a text block to the page. 
2. Add a search block. Notice the placeholder text input has the focus. 
3. Click the text in the text block added from step 1. Notice the focus remains on the text block as expected. 
4. Click the placeholder text. Notice the search block is now selected.
5. Click the button text of the search block and modify it. Verify successful.
6. Click the search label text and modify it. Verify successful. 

## Screenshots <!-- if applicable -->

iOS before - search block stealing focus from text-based blocks

https://user-images.githubusercontent.com/5810477/109748272-96386d00-7ba6-11eb-9c0a-c8bc165318d6.mov

iOS After:

https://user-images.githubusercontent.com/5810477/109748324-ad775a80-7ba6-11eb-8e35-7306ddc8a732.mov

Android before - selecting placeholder text field when block not selected:

https://user-images.githubusercontent.com/5810477/109748446-e31c4380-7ba6-11eb-8fd4-cb7d010309c0.mov



Android after - placeholder text is focused when block first added:

https://user-images.githubusercontent.com/5810477/109748436-df88bc80-7ba6-11eb-9ff1-68e8e73e3b19.mov



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
- Started tracking when different components of the `Search Block` are focused and passing that state via the `isSelected` property of those components. 
- Passed the `Search Block`'s `onFocus` prop to it's child components so when the placeholder text was directly clicked the block would become selected.
- Swapped out the `TextInput` component with the `PlainText` component. 
- Added a `blur()` method to the `PlainText` component and then added a `useEffect` to be called when the`isSelected` property changes to force the `PlainText` to blur when the `Search Block` is no longer selected. Similar things are done in other blocks (`Image`, `MediaText`, `Video`) except they implement a static `getDerivedStateFromProps` method to do the same thing. It works in those blocks because they are React _classes_. My `Search Block` is a _Function Component_ so I'm using `useEffect` to accomplish the same thing. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Commit:
Add search block to inserter

Only adds a shell of a block.

Commit:
Add label and icon toggle toolbar buttons

Commit:
Add button position dropdown menu

Commit:
Add temporary alert to make toolbar testing easier

This alert will be removed once styling is added in a later PR.

Commit:
Add initial UI and basic styling

Commit:
Move text button to separate file

This will keep the code cleaner and easier to follow.

Commit:
Add basic button style

Commit:
Add classnames and consolidate button code

Commit:
Remove iconButtonOnly alert message

This toolbar action is now visible in the UI so the temporary alert message
is no longer needed.

Commit:
Add platform-specific styles and basic styles

Commit:
Remove flex settings for label

Was causing issues with layout in iOS. Styling will be addressed fully
 in a different PR.

Commit:
Add button position styles; remove temp alert

Commit:
Clean up styles

Commit:
Merge branch 'rnmobile/3073-search-block-basic-ui' into rnmobile/3073-search-block-button-position

Commit:
Consolidate border styling, fix auto-sizing

Commit:
Fix broken button style

AmandaRiu committed:
Merge branch 'master' into rnmobile/3073-search-block-basic-ui

Commit:
Merge branch 'rnmobile/3073-search-block-basic-ui' into rnmobile/3073-search-block-button-position

Commit:
Move CSS_UNITs to shared utils file

Moved the CSS_UNIT code to a shared Utils file in preparation for using the same code inside the native implementation of the search block.

Commit:
Add basic settings panel with UnitControl

Commit:
Add event stubs and UnitControl

Commit:
Remove unused code and set defaults

Commit:
Finish changeUnit handler and remove debugging

Commit:
Configure defaults, min and max widths

Also moved the default values from the web version to the shared utils file.

AmandaRiu committed:
Merge branch 'trunk' into rnmobile/3073-search-block-button-position

Commit:
Merge trunk into branch

Commit:
Merge branch 'trunk' into rnmobile/3073-search-block-settings

# Conflicts:
#	packages/block-library/src/search/edit.native.js
#	packages/block-library/src/search/rich-text.android.scss
#	packages/block-library/src/search/rich-text.ios.scss
#	packages/block-library/src/search/style.native.scss

Commit:
Merge branch 'rnmobile/3073-search-block-button-position' into rnmobile/3073-search-block-settings

Commit:
Merge branch 'trunk' into rnmobile/3073-search-block-settings

# Conflicts:
#	packages/block-library/src/search/edit.native.js

Commit:
Fix focus issues on Search Block

- Set the onFocus() method to ensure if the placeholder text is clicked on when the block is NOT selected then the block will be selected.
- Set isSelected in RichText elements to false for initial draw to prevent odd behavior on iOS where the Search block will attempt to steal focus from other text blocks.
- Set the placeholder text to focused when the block is first added.

Commit:
Track the state of which text is selected

Track which of the RichText components are currently selected to avoid focus issues on iOS.

Commit:
Only draw toolbar controls if block is selected

Commit:
Set RichText identifiers and tagNames

Commit:
Clear selection state when the block is unselected

Commit:
Merge branch 'master' into rnmobile/3073-search-block-basic-ui

cameronvoell commented:
Looking much better @AmandaRiu  🙌 

One thing I noticed is that when you _unselect_ the search block while one of the text fields was selected (with the blinking cursor active) - the blinking cursor stays active on that text field even though another block is now selected, and the user can still type in that field even though a new block is selected. I think the expected behavior would be that when you select a new block, the cursor should not remain active in the previously selected text field of the search block. My hunch is that it would be best to address that in this PR, but let me know if you think a separate PR would be better 🙏 

<img src="https://user-images.githubusercontent.com/1103838/110587264-49b1dc00-8128-11eb-9c39-31d707ad3265.jpg" alt="" width="350">

AmandaRiu commented:
Thanks for the review @cameronvoell !

> My hunch is that it would be best to address that in this PR, but let me know if you think a separate PR would be better 

I agree it should be handled in this PR. I'll take a look at it today 🙇‍♀️

Commit:
Make text lose focus when block loses focus

Fixes an issue where if the label or button text was selected in the search block and then a different block was selected, the cursor would still be active in the search block text field.

AmandaRiu commented:
@cameronvoell Issue with text not losing focus when block loses focus has been fixed in 452356ade7ec4d011d9ddf7e561e977eaf7baef3. Ready for another review! 👍

cameronvoell commented:
> @cameronvoell Issue with text not losing focus when block loses focus has been fixed in [452356a](https://github.com/WordPress/gutenberg/commit/452356ade7ec4d011d9ddf7e561e977eaf7baef3). Ready for another review! 👍

@AmandaRiu In my tests it looks like it is fixed for focus on the Button text, and the label text, but not the placeholder text. 

Steps:
1. Add an image block and add an image to it
2. Add a search block
3. select the placeholder text so the cursor is active there and the keyboard is opened.
4. Select the Image block
5. Notice the cursor is still active in the placeholder and the keyboard did not close. Compare this to label and button text where the cursor would disappear and the keyboard would close when you select the image block.

Commit:
Add check for textInput reference existance before manipulation

This isn't really a concern for mobile, but on web there is an option for `button-only` so this
change ensures that we don't try to manipulate a reference that would be null.

Commit:
Rename textInput reference for clarity

Commit:
Ensure textInput loses focus when block loses focus

AmandaRiu commented:
> > @cameronvoell Issue with text not losing focus when block loses focus has been fixed in [452356a](https://github.com/WordPress/gutenberg/commit/452356ade7ec4d011d9ddf7e561e977eaf7baef3). Ready for another review! 👍
> 
> @AmandaRiu In my tests it looks like it is fixed for focus on the Button text, and the label text, but not the placeholder text.
> 
> Steps:
> 
>     1. Add an image block and add an image to it
> 
>     2. Add a search block
> 
>     3. select the placeholder text so the cursor is active there and the keyboard is opened.
> 
>     4. Select the Image block
> 
>     5. Notice the cursor is still active in the placeholder and the keyboard did not close. Compare this to label and button text where the cursor would disappear and the keyboard would close when you select the image block.

Fixed! Ready for another round!

AmandaRiu committed:
Update code comment to remove duplicate wording

Co-authored-by: Paul Von Schrottky <paul.von.schrottky@automattic.com>

AmandaRiu commented on packages/block-library/src/search/edit.native.js:
```diff
@@ -55,6 +73,51 @@ export default function SearchEdit( { attributes, setAttributes, className } ) {
 	} = attributes;
 
+	/*
+	 * Set the focus to the placeholder text when the block is first mounted (if the block
```
Updated. Thanks @guarani

Commit:
Merge branch 'trunk' into rnmobile/search-block-text-focus-bugfixes

Commit:
Add blur method to the PlainText component

Commit:
Use PlainText instead of TextInput for placeholder

AmandaRiu commented:
Hey @guarani  👋  I’ve made the changes we discussed. Ready for another review!

Commit:
Focus the placeholder text when block first added

guarani reviewed (approved):
Tested on iOS and Android, code looks good! Great work, @AmandaRiu! 💯
